src/main/java/com/google/devtools/build/lib/remote/GrpcRemoteOutputService.java

```diff
@@ -0,0 +1,527 @@
+      } catch (StatusRuntimeException e) {
+        throw new IOException(e);
+      }
+      if (responses.getResponsesCount() != 1) {
+        throw new IOException("Remote output service returned an unexpected number of stat responses");
+      }
+      StatResponse response = responses.getResponses(0);
+      if (!response.hasFileStatus()) {
+        // Output service reports that the file does not exist.
+        throw new FileNotFoundException();
+      }
+      FileStatus fileStatus = response.getFileStatus();
+      if (fileStatus.hasFile()) {
+        FileStatus.File regularFileStatus = fileStatus.getFile();
+        return DigestUtil.toBinaryDigest(regularFileStatus.getDigest());
+      }
+      if (fileStatus.hasExternal()) {
+        // Path is a symbolic link that points to a location outside of
+        // the output path.  The output service is unable to give a
+        // digest. We must compute it ourselves.
+        return super.getFastDigest(
+            originalOutputPath.getRelative(fileStatus.getExternal().getNextPath()));
+      }
+      throw new IOException(format("Remote output service did not return the status of file %s", path));
+    }
+  }
+
+  @Override
+  @Nullable
+  public FileSystem createActionFileSystem(
+      FileSystem sourceDelegate,
+      PathFragment execRootFragment,
+      String relativeOutputPath,
+      ImmutableList<Root> sourceRoots,
+      ActionInputMap inputArtifactData,
+      Iterable<Artifact> outputArtifacts,
+      boolean trackFailedRemoteReads) {
+    return new GrpcFileSystem(sourceDelegate, execRootFragment.getRelative(relativeOutputPath));
+  }
+
+  @Override
+  public void checkActionFileSystemForLostInputs(FileSystem actionFileSystem, Action action)
+      throws LostInputsActionExecutionException {
+    // There is no need to provide an implementation for this function,
+    // as StartBuild() on the server is supposed to call
+    // FindMissingBlobs() for all remotely tracked files. This ensures
+    // that they cannot disappear during the build.
+  }
+
+  public ListenableFuture<Void> downloadActionResult(ActionResult actionResult) {
+    // Request that all outputs of the action are created. Do make sure
+    // to remove the "bazel-out/" prefix from all paths, as that part of
+    // the path is not managed by the remote output service.
+    BatchCreateRequest.Builder builder = BatchCreateRequest.newBuilder();
+    builder.setBuildId(stateAcrossBuilds.currentBuildId.toString());
+    for (OutputFile file : actionResult.getOutputFilesList()) {
+      // As there is no guarantee that permissions on files created by
+      // the remote output service can be modified, make sure that all
+      // downloaded files are marked executable.
+      // TODO: Do we only want to do this selectively?
+      builder.addFilesBuilder()
+          .mergeFrom(file)
+          .setPath(fixupExecRootPath(PathFragment.create(file.getPath())))
+          .setIsExecutable(true);
```

Comment:
Do we ever insert a file which is not marked as executable? There is a request to have other modes supported, but we can maybe defer adding support for that until we get to #5588?